src/Facets/HopFacetPacked.sol

```diff
@@ -0,0 +1,435 @@
+pragma solidity 0.8.17;
+
+import { IHopBridge } from "../Interfaces/IHopBridge.sol";
+import { ILiFi } from "../Interfaces/ILiFi.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/// @title Hop Facet (Optimized for Rollups)
+/// @author LI.FI (https://li.fi)
+/// @notice Provides functionality for bridging through Hop
+contract HopFacetPacked is ILiFi {
+
+    /// External Methods ///
+
+    /// @notice Bridges Native tokens via Hop Protocol from L2
+    /// No params, all data will be extracted from manually encoded callData
+    function startBridgeTokensViaHopL2NativePacked(
+    ) external payable {
+        checkCalldataLength(120);
+        _startBridgeTokensViaHopL2Native({
+            // first 4 bytes are function signature
+            transactionId: bytes32(getCalldataValue(4, 8)), // bytes8 > bytes32
+            integrator: string(abi.encodePacked(getCalldataValue(12, 16))), // bytes16 > string
+            receiver: address(uint160(getCalldataValue(28, 20))), // bytes20 > address
+            destinationChainId: getCalldataValue(48, 4), // bytes4(uint32) > uint256
+            bonderFee: getCalldataValue(52, 16), // bytes16(uint128) > uint256
+            amountOutMin: getCalldataValue(68, 16), // bytes16(uint128) > uint256
+            destinationAmountOutMin: getCalldataValue(84, 16), // bytes16(uint128) > uint256
+            hopBridge: address(uint160(getCalldataValue(100, 20))) // bytes20 > address
+            // => total calldata length required: 120
+        });
+    }
```

Comment:
An alternative way to decode using only solidity:
```solidity
(bytes8 transactionId,
bytes16 integrator,
address receiver,
bytes4 destinationChainId,
bytes16 bonderFee,
bytes16 amountOutMin,
bytes16 destinationAmountOutMin,
address hopBridge) 
= abi.decode(msg.data[4:], 
(bytes8,
bytes16,
address,
bytes4,
bytes16,
bytes16,
bytes16,
address));
```

Reply:
@maxklenk We can avoid low-level assembly with this succinct method.

I'm okay with both approaches. Solidity would revert if there are extra call data to decode.
Use `abi.decode(msg.data[4:120]` if we want to attach extra information in the calldata. https://github.com/spearbit-audits/review-lifi-retainer1/issues/45#issuecomment-1485414924

Reply:
Doesn't seem to work after all

Reply:
@gpersoon it would be great to be able to decode in solidity, but I was not able to get the suggested `abi.decode` function to work. As `abi.decode` is the counterpart to `abi.encode` which encodes all values in `bytes32` length, I assume the the decoding also expects that the data is formatted that way.